### ISSUE TITLE
Add an `Integer` generation method

### ISSUE THREAD

**Issue by rlouf** [normal-computing/outlines#153]
We can steer the generation process so we only output integers. I propose to implement an `Integer` subclass of `Sequence` that uses [masking](https://github.com/normal-computing/outlines/blob/a4662218232df8264d1b9eb93867362ed8da2e54/outlines/text/masks.py#L36) to restrict the generated tokens to integers. We will need to add a `create_proposal` method to `Sequence` that applies the mask to the logits generated by the model.

We can also add constraints on the generated integers:

* Add an upper or lower bound on the number of digits generated
* Add an upper or lower bound on the value of the integer

We will probably need [SMC sampling](https://arxiv.org/abs/2306.03081v1) to implement the constraints.
